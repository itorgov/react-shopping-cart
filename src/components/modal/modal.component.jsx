import React, {useEffect, useState} from "react";
import ReactDOM from 'react-dom';

import {Button} from "../button/button.component";

import styles from './modal.module.scss';

export const Modal = ({
        children: content,
        buttonVariant = 'black',
        buttonSmall = false,
        isOpened = false,
        onOpened,
        onClosed
    }) => {
    const [opened, setOpened] = useState(isOpened);
    const el = document.createElement('div');
    const modalRoot = document.getElementById('modal-root');

    useEffect(() => {
        modalRoot.appendChild(el);

        return () => {
            modalRoot.removeChild(el);
        };
    });

    const openModal = () => {
        setOpened(true);

        if (onOpened) {
            onOpened();
        }
    }

    const closeModal = () => {
        setOpened(false);

        if (onClosed) {
            onClosed();
        }
    }

    return (
        <>
            <Button
                variant={buttonVariant}
                small={buttonSmall}
                onClick={() => openModal()}
            >
                Add more
            </Button>
            {opened ? ReactDOM.createPortal(
                <>
                    <div
                        className={styles.modal}
                        onClick={event => event.stopPropagation()}
                    >
                        <div className={styles.header}>
                            <h2 className={styles.title}>
                                Choose a product
                            </h2>
                            <button
                                className={styles.close}
                                onClick={() => closeModal()}
                            >
                                &times;
                            </button>
                        </div>
                        <div className={styles.content}>
                            {content}
                        </div>
                    </div>
                    <div
                        className={styles.cover}
                        onClick={() => closeModal()}
                    />
                </>,
                el
            ) : ''}
        </>
    );
}
