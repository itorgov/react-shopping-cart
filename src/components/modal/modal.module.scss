.cover {
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  background-color: rgba(var(--color-black-rgb), 0.4);
  z-index: 1000;
}

.modal {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 80vw;
  max-width: 114rem;
  max-height: 96%;
  background-color: var(--color-white);
  border-radius: 3px;
  box-shadow: 0 3rem 6rem rgba(var(--color-secondary-rgb), 0.2);
  z-index: 1010;
  overflow: auto;

  @media only screen and (max-width: 43.75em) {
    width: 96vw;
  }

  @media only screen and (max-width: 37.5em) {
    min-height: 100%;
    max-height: 100%;
    min-width: 100%;
    max-width: 100%;
    border-radius: 0;
  }

  .header {
    display: flex;

    .title {
      flex: 1;
      padding: 1rem 2rem;
      color: var(--color-text-dark);
      font-size: 2.5rem;
      font-weight: 300;
    }

    .close {
      display: inline-block;
      padding: 0 2rem;
      font-size: 3rem;
      border: 0;
      border-radius: 0;
      background-color: transparent;
      color: var(--color-text-light);
      transition: all 0.2s;

      &:focus {
        outline: none;
      }

      &:hover,
      &:active,
      &:focus {
        color: var(--color-text-dark);
      }
    }
  }

  .content {
    padding: 1rem 2rem;
  }
}
