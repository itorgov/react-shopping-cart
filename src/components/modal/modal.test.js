import React from "react";
import {shallow} from "enzyme";

import {Modal} from './modal.component';

describe('<Modal />', () => {
    it('expect to render button as default', () => {
        expect(shallow(<Modal />)).toMatchSnapshot();
    });
});

describe('<Modal />', () => {
    it('expect to render modal when it has opened prop', () => {
        expect(shallow(<Modal isOpened={true} />)).toMatchSnapshot();
    });
});
