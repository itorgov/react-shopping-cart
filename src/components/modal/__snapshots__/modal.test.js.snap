// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> expect to render button as default 1`] = `
<Fragment>
  <Button
    onClick={[Function]}
    small={false}
    variant="black"
  >
    Add more
  </Button>
</Fragment>
`;

exports[`<Modal /> expect to render modal when it has opened prop 1`] = `
<Fragment>
  <Button
    onClick={[Function]}
    small={false}
    variant="black"
  >
    Add more
  </Button>
  <Portal
    containerInfo={<div />}
  >
    <div
      className="modal"
      onClick={[Function]}
    >
      <div
        className="header"
      >
        <h2
          className="title"
        >
          Choose a product
        </h2>
        <button
          className="close"
          onClick={[Function]}
        >
          Ã—
        </button>
      </div>
      <div
        className="content"
      />
    </div>
    <div
      className="cover"
      onClick={[Function]}
    />
  </Portal>
</Fragment>
`;
